### PR TITLE
[wasm][debugger] Support array expansion in VS via callFunctionOn

### LINKED ISSUE #19229: Unable to inspect local array values when debugging a Blazor WebAssembly app
@lewing @SteveSandersonMS 

Repro steps:

- Use latest VS dogfood build (16.6.0 Preview 2.0 [29913.155.master]) and latest internal Blazor WebAssemby 3.2 Preview 3 build (3.2.0-preview3.20163.13)
- Create an ASP.NET Core hosted Blazor WebAssembly app
- In the Client project set a breakpoint in `FetchData.OnInitializedAsync` on the line *after* the HTTP request
- Hit F5 to run the app in the debugger
- Navigate to the FetchData tab
- The breakpoint should hit
- Inspect the forecasts array local variable

Expected results:
- See the expanded elements of the array

Actual result:
- No elements of the array are visible:

image

### LINKED ISSUE #19531: [wasm][debugger] Support `Runtime.callFunctionOn`
This is used by vscode js debugger, and chrome debugger to expand arrays.
This would also allow big arrays to be expanded correctly.

related issue: https://github.com/mono/mono/issues/19392

### PR DESCRIPTION
**Pull request by radical**
- This was prompted by vscode/vs requiring this for array expansion.
- It runs a given js function string, on a given object id.
- We handle the case for dotnet object ids.

Implementation:

- We build a proxy object that reflects the details of the real dotnet object/array
- And run the js function on that proxy object

- Then if `returnByValue` was requested, the result is returned as-is
- Else we cache that object, with a new special object id (`dotnet:cfo_res:..`),
  and return that id in the response.

- Subsequently, we would get `Runtime.getProperties` request on this new object id
- And a `Runtime.releaseObject` request, to free that function result.

- All this is handled currently

Tests:

- A few tests were added which invoke `callFunctionOn`, on JS objects, and checks the result
  - And the same checks are performed on equivalent dotnet objects, with the same js function
  - This should help stay in sync with what we think `callFunctionOn` should work as

  - These tests have an additional parameter - `roundtrip`, which calls a simple function
	`function () { return this; }`
    .. on a dotnet object, and gets the result object id. If `roundtrip == true`, then it
    runs that same function on the result object, and gets a new result object id. And the
    tests run their checks on that.

    - this helps to check that outside of the proxy, the result object behaves same as any
      other dotnet object.

- Also, most of the tests were modified to support an additional mode (`use_cfo`).
  - With this, whenever a test wants to run `Runtime.getProperties on an object, we pass that
    object through the earlier "identity" function, and then run `getProperties` on that
    resultant object.

  - This helps to take advantage most of the existing tests, and use them for `callFunctionOn`
    testing too, by ensuring that the results of CFO returns results same as any other object.

Fixes #19229 
Fixes #19531

### CODE REVIEW

**Comment by lewing** [comment]
@monojenkins build failed

File: sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs [hunk 1]
```diff
@@ -216,38 +216,34 @@ protected override async Task<bool> AcceptCommand (MessageId id, string method,
...
-							}
-						}
+					if (!objId.StartsWith ("dotnet:", StringComparison.Ordinal))
+						// not for us
+						break;
 
+					var parts = objId.Split (new char [] { ':' });
+					if (parts.Length < 3) {
+						// FIXME: um return an error?
 						return true;
```
Comment (lewing) [inline sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs:226]:
```suggestion
						return false;
```
Can't return true without sending a response

File: sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs [hunk 2]
```diff
@@ -216,38 +216,34 @@ protected override async Task<bool> AcceptCommand (MessageId id, string method,
...
 					}
+
+					var result = await RuntimeGetProperties (id, objId, parts, args, token);
+					if (result.IsErr) {
+						SendResponse (id, result, token);
+						return true;
+					}
+
+					SendResponse (id, result, token);
+					return true;
```
Comment (lewing) [inline sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs:236]:
```suggestion
					SendResponse (id, result, token);
					return true;
```
These do the same thing

**Comment by lewing** [comment]
we're going to need the stuff in https://github.com/mono/mono/pull/19590 as well

**Comment by lewing** [comment]
@monojenkins build failed

File: sdks/wasm/src/library_mono.js [hunk 3]
```diff
@@ -229,18 +230,242 @@ var MonoSupportLib = {
...
+		_cache_call_function_res: function (obj) {
+			var id = `dotnet:cfo_res:${this._next_call_function_res_id++}`;
+			this._call_function_res_cache[id] = obj;
+			return id;
+		},
+
+		_x: function (obj, n) {
+			obj == undefined ? "" : JSON.stringify(obj, undefined, 4).slice(0, n);
+		},
+
```
Comment (lewing) [inline sdks/wasm/src/library_mono.js:393]:
```suggestion
```
This doesn't appear to be used?

File: sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs [hunk 4]
```diff
@@ -216,38 +216,29 @@ protected override async Task<bool> AcceptCommand (MessageId id, string method,
...
 
+					var result = await RuntimeGetProperties (id, objId, parts, args, token);
+					SendResponse (id, result, token);
+					return true;
+				}
+
+			case "Runtime.releaseObject": {
+					var objId = args ["objectId"]?.Value<string> ();
+					if (objId?.StartsWith ("dotnet:cfo_res:", StringComparison.Ordinal) == true) {
+						await SendMonoCommand (id, new MonoCommands ($"MONO.mono_wasm_release_object('{objId}')"), token);
```
Comment (lewing) [inline sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs:237]: The idea behind the admittedly confusing static MonoCommands was a level of type safety for the arguments, is there a reason you aren't adding the commands there?

Comment (radical) [reply sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs:237]: Oops, I missed that!

**Comment by lewing** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

File: sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs (hunk 4 above)
Comment (lewing) [reply sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs:237]: Easy to fix in a follow up.

**Comment by lewing** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

**Comment by lewing** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

**Comment by lewing** [comment]
Failing because of https://github.com/mono/mono/pull/19604

**Comment by lewing** [comment]
@monojenkins backport 3.2

**Comment by lewing** [comment]
@monojenkins backport 3.2-wasm

**Pull request merged**